### PR TITLE
Update Cypress plugin to reflect new cross-browser capabilities

### PR DESCRIPTION
**Pull request by amirrustam**
Cypress version 4.0.0 will be introducing Firefox and MS Edge support. Vue CLI Cypress plugin users will now be able to run tests in Chromium-based browsers, Firefox, and Electron. This PR updates the CLI messaging and plugin readme to reflect these upcoming changes.

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [X] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Docs
- [ ] Underlying tools
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by vesper8** [comment]
The migration guide mentions at least 5 breaking changes to look out for when upgrading to 4.0 so I'm not sure if it's accurate to say that changing the dependency of Cypress to ^4.0 causes no breaking changes

I certainly see this breaking unaware people's cypress tests in many cases

https://docs.cypress.io/guides/references/migration-guide.html#Migrating-to-Cypress-4-0

**Comment by amirrustam** [comment]
Thanks for the feedback @vesper8.

Breaking changes are rare with Cypress, but when they are introduced during a major version bump, we take great care to communicate the changes to users and guide them to resources/examples for a quick resolution. 

For example, let's take a look at the [first migration example](https://docs.cypress.io/guides/references/migration-guide.html#Example-1) from the guide you linked regarding the use of a `done()` callback while a test returns a promise. I can simulate this scenario with the following:

```js
describe('4.0 Migration Scenarios', () => {
  it('mocha upgrade', (done) => {
    cy.visit('https://example.cypress.io')

    return new Promise((resolve) => {
      resolve()
    }).then(() => {
      done()
    })
  })
})
```

When we run this test, it does not mysteriously fail and leave the user confused. Instead, the user will see the following:

Screen Shot 2020-02-11 at 10 14 27 AM

1. The error is explained in plain English.
2. Link user to related fix, guide, or explanation.

We actually have a page in the Cypress docs that [explains all error message](https://docs.cypress.io/guides/references/error-messages.html), and we link to each one as appropriate from Cypress.

The important thing is that we're upfront and helpful to the user during these moments.

Some questions to move things forward:
- Will Cypress breaking changes require a major release for Vue CLI?
- Is the breaking change communication/experience of Cypress sufficient for Vue CLI users after an upgrade?

**Comment by vesper8** [comment]
Thank you for your helpful message @amirrustam, to be clear I've already upgraded to Cypress 4 and found the experience easy and smooth. I for one would certainly love for Cypress 4 to become the new default in vue-cli v4. I was simply pointing out that the upgrade does mention a few breaking changes and this will in all likelihood result in tests breaking so the answer in the initial post indicating that this PR brings in no breaking change was questionable. I sure do appreciate the usefulness of the warnings and errors that Cypress can return.

**Comment by Akryum** [comment]
I don't think we can push a minor or a patch update that breaks our users tests even if Cypress can document itself. Also, we have enterprise users that are not as flexible as solo developers and it would be a hassle to require them to update their test in a "non-breaking" update. We would need to bump Vue CLI version to 5.0 imo.

**Comment by Akryum** [comment]
@sodatea Do we have a 5.0 release planned already?

**Comment by sodatea** [comment]
I've created a project board for v5.

But I'd more like to wait until Webpack 5 reaches RC phase to start working on it.

File: packages/@vue/cli-plugin-e2e-cypress/package.json [hunk 1]
```diff
@@ -24,7 +24,7 @@
   },
   "dependencies": {
     "@vue/cli-shared-utils": "^4.2.2",
-    "cypress": "^3.8.3",
+    "cypress": "^4.0.0",
```
Comment (loter) [inline packages/@vue/cli-plugin-e2e-cypress/package.json]: Should be 4.0.1

Comment (amirrustam) [reply packages/@vue/cli-plugin-e2e-cypress/package.json]: we released a quick release patch, will update. 👍

**Comment by amirrustam** [comment]
@sodatea thank you for including this within your v5 plans. Is there a timeframe for Webpack 5 RC release?

It would be great to deliver the new Firefox support and other Cypress v4 benefits to Vue CLI users sooner than later. I appreciate your consideration.

**Comment by jacobbundren** [comment]
Is this going into an upcoming release?

File: packages/@vue/cli-plugin-e2e-cypress/package.json [hunk 2]
```diff
@@ -24,7 +24,7 @@
   },
   "dependencies": {
     "@vue/cli-shared-utils": "^4.2.2",
-    "cypress": "^3.8.3",
+    "cypress": "^4.0.1",
```
Comment (MickaelBergem) [inline packages/@vue/cli-plugin-e2e-cypress/package.json:27]: FWIW the latest version is now `4.3.0` :)

**Comment by amirrustam** [comment]
@jacobbundren This will be added in the v5 version of the CLI.

@MickaelBergem There will be many more releases of Cypress by the time v5 of the CLI is released. I'll update the PR with the latest version prior to the CLI release.

**Comment by doits** [comment]
FYI: Until a new vue-cli version is released you can manually force the latest version if you use `yarn` with `resolutions`. In your `package.json`:

```json
{
  "resolutions": {
    "cypress": "^4.4.0"
  }
}
```

**Comment by ZwapKillrath** [comment]
Any news yet...Waiting eagerly :-)

**Comment by igorshubovych** [comment]
It make sense to upgrade to Cypress at least 4.4.0, because it better supports TS.

**Comment by sodatea** [comment]
My updated plan:
Ship Vue CLI 4.5 next month with preliminary 1st-party support for Vue 3.
Then start working on the next major version regardless of the progress of webpack 5.

File: packages/@vue/cli-plugin-e2e-cypress/package.json (hunk 2 above)
Comment (smarandav) [reply packages/@vue/cli-plugin-e2e-cypress/package.json:27]: Can we use a  Cypress >=4.5.0 as there is a dependency down the chain on cypress on mocha 2.4.5 that is reeeay old and causes issues with other plugins? Many thanks!

And any luck for this pr to be merged soon? :)

**Comment by smarandav** [comment]
Any news when this is going to be merged as we are eagerly waiting for the bump of cypress version above 4.5, thanks!! @sodatea

**Comment by tschoartschi** [comment]
@smarandav we are also waiting for the newest version of Cypress. In the meanwhile we use the following workaround: https://github.com/vuejs/vue-cli/issues/5352

We bumped to `4.8.0` and we are not experiencing any troubles.

File: packages/@vue/cli-plugin-e2e-cypress/package.json (hunk 2 above)
Comment (upcccz) [reply packages/@vue/cli-plugin-e2e-cypress/package.json:27]: why not use 4.4.0+ to support TS?

**Comment by tongvantruong** [comment]
Hi bros, I am waiting for this change. Thanks

**Comment by maltewirz** [comment]
Hi there, also waiting for this. Many thanks!

**Comment by mathew-ferry13** [comment]
@yyx990803 Any updates?

**Comment by sodatea** [comment]
Scheduled after the v4.5 release.

For v4.5, I'm still reviewing https://github.com/vuejs/vue-cli/pull/5356 https://github.com/vuejs/vue-cli/pull/5479 and working on https://github.com/vuejs/vue-cli/pull/5637

**Comment by amirrustam** [comment]
We'll be releasing Cypress version 5.0 soon. I'll update this PR after the release, so CLI 4.5 can at least be anchored to the next major version of Cypress.

**Comment by layanto** [comment]
@amirrustam Will component unit testing via cypress be supported in addition to the currently supported unit testing with Mocha+Chai or Jest? See https://github.com/JessicaSachs/vue-cli-plugin-cypress-experimental

**Comment by ZwapKillrath** [comment]
> We'll be releasing Cypress version 5.0 soon. I'll update this PR after the release, so CLI 4.5 can at least be anchored to the next major version of Cypress.

Hi Amir,

When do you expect 5.0 to be released?
(no pressure added :-))

**Comment by Mikilll94** [comment]
Any updates on this?

**Comment by amirrustam** [comment]
We just released Cypress v5.0. Updating this PR now.

**Comment by amirrustam** [comment]
@sodatea looks like the previous version of this PR was slated for 4.5, but now it's scheduled for 5.0. What's the expected time frame for CLI v5?

We'd like Cypress version updates to be easier in the future. So are you open to this idea? https://github.com/vuejs/vue-cli/issues/5703#issuecomment-668073623

Side-topic: my latest commit failed a Cypress test, I can easily resolve this locally, but a faster way to see what went wrong is the [Cypress Dashboard](https://dashboard.cypress.io/). I can get Vue CLI project OSS plan access (essentially full-featured for free), would you be interested?

**Comment by sodatea** [comment]
@amirrustam 
I'm working on the development plan of v5 this week.

According to past experience, it may cost ~3 months before the stable release. An alpha version may come out before the end of October.

And yeah, Cypress should be a peer dependency in the next major. It's a regret that we didn't make it in v4.

**Comment by sodatea** [comment]
The Cypress Dashboard looks really cool! I'd love to have access to its features. Thanks very much 😄

**Comment by sodatea** [comment]
Merged via command line with https://github.com/vuejs/vue-cli/commit/3717f7c973f2c1950e2b1b43c778071dc1ff8671 & https://github.com/vuejs/vue-cli/commit/149e3f1fa60b39516237fc46271fb2829b383781
